pkg/router/template/router.go

```diff
@@ -311,31 +323,171 @@ func (r *templateRouter) readState() error {
+		glog.V(8).Infof("Commit called and the state has changed, could reload")
+
+		if r.commitReqTime == nil {
+			// There is no scheduled commit worker, so set the time we started and
+			// invoke the worker code.  It will decide if it can run now, or if it
+			// needs to schedule a callback.
+			//
+			// We need to track the earliest commit time so we can do burst supression.
+			now := time.Now()
+			r.commitReqTime = &now
+			r.lock.Unlock()
+
+			glog.V(8).Infof("No scheduled reload, calling the worker (curtime: %v)", now)
+
+			r.commitWorker()
+			return
+		}
+
+		glog.V(8).Infof("There is already a scheduled reload (for %v), skipping the worker", r.commitReqTime)
+	}
+
+	r.lock.Unlock()
+}
+
+// timeUntilNextAction() works out when we can next reload based on the current time, the last action time, and the minimum allowed gap between the two.
+// In order to be allowed to reload immediately, the last action + minumum gap must be < the current time.
+// If we can reload, then it returns a zero duration.
+// If we can't reload, then it reuturns a duration to wait for before a reload would be allowed.
+func timeUntilNextAction(now time.Time, lastActionTime time.Time, minimumActionGap time.Duration) (nextReload time.Duration) {
+	if sinceLastAction := now.Sub(lastActionTime); sinceLastAction < minimumActionGap {
+		return minimumActionGap - sinceLastAction
 	}
 
-	needsCommit := r.stateChanged
+	return 0 * time.Second
+}
+
+func (r *templateRouter) commitWorker() {
+	r.lock.Lock()
+
+	glog.V(8).Infof("CommitWorker called")
+
+	if r.commitRunning {
+		// We don't need to do anything else... there's a commit in progress, and when it is done it will re-call this function at which point the work will then happen
+		glog.V(8).Infof("There was already a commit running (%v) returning from the worker", r.commitRunning)
+		r.lock.Unlock()
+		return
+	}
+
+	if r.commitReqTime == nil {
+		// There's no commit queued so we have nothing to do.  We should only get here when
+		// the function is re-called after a reload
+		glog.V(8).Infof("No commit requested time, so there's no queued commit.  Nothing to do.")
+		r.lock.Unlock()
+		return
+	}
+
+	// There is no commit running, let's see if we should run yet, or schedule a callback
+	var untilNextCallback time.Duration
+	now := time.Now()
+
+	getNextReload := func(potentialDuration time.Duration) time.Duration {
+		// Use the largest of the potential durations so we only come back when the checks will allow it to run
+		if potentialDuration > untilNextCallback {
+			return potentialDuration
+		}
+
+		return untilNextCallback
+	}
+
+	untilNextCallback = getNextReload(timeUntilNextAction(now, r.lastReloadStart, r.reloadInterval))
+	untilNextCallback = getNextReload(timeUntilNextAction(now, r.lastReloadEnd, r.reloadGap))
+	untilNextCallback = getNextReload(timeUntilNextAction(now, *r.commitReqTime, r.reloadEventWait))
+
+	if untilNextCallback > 0*time.Second {
```

Comment:
you don't need `* time.Second` here, I believe

Reply:
Done